designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+
+### Declaring Variables with User-defined Constraints
+
+Variables can be declared with user-defined constraints in the data, transformed data, parameters, transformed parameters, and generated quantities blocks. The syntax for declaring a variable with a user-defined constraint is as follows:
+
+```stan
+type<udc=constraint_name(args...)> variable_name;
+```
+
+Here, udc denotes a user-defined constraint , and constraint_name(args...) specifies the name of the constraint along with any arguments it requires.
+
+### Example Usage
+
+```stan
+constraints {
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real constrain(vector x, real upper_bound) {
+     jacobian += sum(x);
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+data {
+ real a;
+ real u_bound;
+ int N;
+}
+
+
+transformed data {
+ // Data will use the validate() function from the user-defined constraint
+ real<udc=upper_bound(u_bound)> a_validated;
+}
+
+
+parameters {
+ // User-defined constraint applied to parameters alpha and b
+ real<udc=upper_bound(u_bound)> alpha;
+ vector<udc=upper_bound(u_bound)> b;
+ // In these cases this is the same as calling
+ // real<upper=u_bound> alpha;
+ // vector<upper=u_bound>[N] b;
+}
+```
+
+In this example, upper_bound is a user-defined constraint that is applied to the data and parameters. The validate() function from upper_bound is used to validate a_validated in the transformed data block, and the transformations defined in upper_bound are applied to alpha and b in the parameters block.
+
+### Error Messages
+
+If the `constrain` function of a constraint does not increment `jacobian` a warning is thrown to inform users. In these cases it is better to use a standard function as a new constraint type is most likely not necessary.
+
+```md
+Warning: {CONSTRAINT_TYPE} does not increment the log of the absolute of the determinant of the jacobian.
+It is recommended to instead use a function in `transformed parameters` instead of a custom constraint.
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The first thing we would do is add the `jacobian` keyword to the Stan language. Like `target`, `jacobian` would only be available on the left hand side of statements to be accumulated into. Its domain is restricted to the `transformed parameters` and `constraints` block.
```

Comment:
The motivation is missing that we'd like to be able to implement transforms in Stan that match the ones supplied by Stan.  So the key thing to mention here is that `jacobian` acts differently than `target` because it will get turned off in maximum likelihood estimation.

Reply:
👍